src/software/geom/polynomial.cpp

```diff
@@ -0,0 +1,50 @@
+#include "software/geom/polynomial.h"
+
+Polynomial::Polynomial(const std::vector<double>& coeffs) : coeffs(coeffs)
+{
+    pruneCoeffs();
+}
+
+Polynomial::Polynomial(const std::initializer_list<double>& coeffs) : coeffs(coeffs)
+{
+    pruneCoeffs();
+}
+
+Polynomial::Polynomial(const std::pair<double, double>& p1,
+                       const std::pair<double, double>& p2)
```

Comment:
Why don't we use our `point` class? Also what if I give `{1, 0}` and `{1, 2}`? Won't we segfault? I don't think it really makes any sense to construct a polynomial from two points. If we want a 2D line segment, then that should be two polynomials (thinking back to how we break a spline up into two functions `f_x(t)` and `f_y(t)`).

Reply:
it's not a point... it's a map to some input (double) to some output (double). Polynomial could be used to map point.x to point.y, but that won't always be the case

Reply:
this polynomial takes any two input/output pairs and gives you a polynomial that represents a line between the input/output pairs

Reply:
spline does use two polynomials to represent a 2D line segment. This is polynomial, not 2D line segment

Reply:
It seems like the implementation treats the pairs of `double` semantically as points?
>Polynomial could be used to map point.x to point.y, but that won't always be the case

I don't totally understand what you mean here because these are 2 pairs of values, not just a mapping one one value to another?

Reply:
This probably super confusing because they look like points. Say we have a polynomial to map some value q to some value r. Then we give the constructor two pair of input/output (q1,r1) and (q2, r2). The polynomial constructed must return r1 if q1 is inputted and r2 is q2 is inputted. (q1,r1) and (q2, r2) are represented by p1, p2. To Gareth's point, if q1==q2, then we have a problem...

Reply:
That makes sense, and using pairs of doubles is good since it makes this a more generic class...

But considering we just take the two pairs of input/output and create a degree-1 polynomial between them, seems like it's not different than 2 points in space? Or is this going to be generalized to higher degrees / more pairs later?

Reply:
Yes it should be expanded to higher degrees. It should probably be a factory at that point. Polynomial class maybe is a math class and shouldn't know about points or lines. It's just a function with input/output. Tying it to Point seems to constrain Polynomial too much.

Reply:
That seems reasonable

Reply:
I don't think this is a reasonable signature because you can easily pass clearly invalid data to it (ex. if `p2.first` and `p1.first` are equal). If we need to construct a first order polynomial from two points, we should take a point and a slope, and make it the responsibility of the caller (who has the semantic context) to not divide by zero. Also this probably shouldn't be a class member as you say.

Reply:
done

Reply:
Sure